cep-purls.md

```diff
@@ -0,0 +1,192 @@
+
+## Specification
+
+We propose to add the optional `purls: [string]` field to `PackageRecord`.
+To identify the repackaged package we use [PURL](https://github.com/package-url/purl-spec/) (Package URL), which implements a scheme for identifying packages that is meant to be portable across packaging ecosystems.
+
+```json
+{
+  ...
+  "pinject-0.14.1-pyhd8ed1ab_0.tar.bz2": {
+    "name": "pinject",
+    "version": "0.14.1",
+    "purls": ["pkg:pypi/pinject@0.14.1"],
+    ...
+  }
+  ...
+}
+```
+
+PURL is already supported by dependency-related tooling like SPDX (see [External Repository Identifiers in the SPDX 2.3 spec](https://spdx.github.io/spdx-spec/v2.3/external-repository-identifiers/#f35-purl)), the [Open Source Vulnerability format](https://ossf.github.io/osv-schema/#affectedpackage-field), and the [Sonatype OSS Index](https://ossindex.sonatype.org/doc/coordinates); not having to wait years before support in such tooling arrives is valuable.
+[PEP 725 (WIP)](https://peps.python.org/pep-0725) also proposes how to specify non-PyPi dependencies using PURLs.
+
+### PURLs in recipes
+
+PURLs can be specified in the recipe under `.about.purls` (implementation for [v1-recipes](cep-00014.md): [prefix-dev/rattler-build #1664](https://github.com/prefix-dev/rattler-build/pull/1664)):
+
+```yaml
+about:
+  # ...
+  # PURLs that this package belongs to
+  purls: [string (PURL enforced)]
+```
+
+Tools that build conda packages should add these packages to `index.json`:
+
+```json
+{
+  ...,
+  "purls": ["pkg:pypi/pinject@0.14.1"]
+}
+```
+
+### PURLs in repodata
+
+In `repodata.json`, PURLs should be an optional field.
+Tools that generate `repodata.json` (like `rattler-index` or `conda-index`) can decide whether to include it or not in `repodata.json`.
+
+Instead of writing the PURLs in `repodata.json`, these tools can decide to add PURLs to a separate `purls.json` like in [CEP 12](cep-0012.md):
+
+```json
+{
+  "info": {
+    "platform": "string",
+    "arch": "string",
+    "subdir": "string",
+    "version": 0
+  },
+  "packages": {
+    "package-version-build.tar.bz2": {
+      ...,
+      "purls": ["pkg:pypi/pinject@0.14.1"]
+    }
+  },
+  "packages.conda": {
+    "package-version-build.conda": {
+      ...,
+      "purls": ["pkg:pypi/pinject@0.14.1"]
+    }
+  }
+}
+```
+
+If the `purls` field is not present in the `repodata.json` record, it means no `purls` information is stored with the record. Then, a fallback mechanism (i.e., read `purls.json`) should be used to acquire the PURL information.
+
+In sharded repodata ([CEP 16](cep-0016.md)), PURLs can be specified similar to [CEP 21](cep-0021.md).
+
+### Patching
+
+With the purls part of the repodata we propose to also allow repodata patching this field. The original purls can still be extracted from the packages in `index.json`. We do not see a reason how the PURL information is different from other patchable information stored in the repodata (like the dependencies).
```

Comment:
```suggestion
With the purls part of the repodata we propose to also allow repodata patching this field. The original purls can still be extracted from the packages in `index.json`. We do not see a reason how the PURL information is different from other patchable information stored in the repodata (like the dependencies).
```

I'm not sure we want to dynamically alter a field that might end up being used to generate SBOMs and other types of audit-related data. I need to think about this.